### ISSUE TITLE
Docker does not conform to the Docker registry spec wrt token auth

### ISSUE THREAD

**Issue by majewsky** [docker/distribution#2660]
```

**Comment by manishtomar** [comment]
I believe dockerd is trying to verify if your registry supports v2 protocol by doing a simple GET on /v2/ as per https://docs.docker.com/registry/spec/api/#api-version-check. Since it is not receiving a 200 it is falling back to v1 protocol which has been deprecated long time ago. I would suggest returning 200 on `GET /v2/` and seeing how it behaves.

**Comment by majewsky** [comment]
Are you sure? By that standard, Docker Hub does not support the v2 API:

```
$ curl -i https://registry-1.docker.io/v2/
HTTP/1.1 401 Unauthorized
Content-Type: application/json; charset=utf-8
Docker-Distribution-Api-Version: registry/2.0
Www-Authenticate: Bearer realm="https://auth.docker.io/token",service="registry.docker.io"
Date: Tue, 07 Aug 2018 08:39:06 GMT
Content-Length: 87
Strict-Transport-Security: max-age=31536000

{"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":null}]}
```

**Comment by manishtomar** [comment]
I believe dockerd checks the `Docker-Distribution-API-Version` response header as per following statement from the spec:
> When a 200 OK or 401 Unauthorized response is returned, the “Docker-Distribution-API-Version” header should be set to “registry/2.0”. Clients may require this header value to determine if the endpoint serves this API. When this header is omitted, clients may fallback to an older API version.

**Comment by majewsky** [comment]
I thought about this, and [added the header](https://github.com/sapcc/keppel/commit/2047c5c1b07c55178aa7a859afafc9c2fbcd4833) to my 401 response, but it doesn't help. Also, as I said above, Docker behaves the same when pushing to Docker Hub, which also returns the API version header on all responses (incl. 401).

**Comment by manishtomar** [comment]
Hmm. I am not sure why it is not working after setting the header. I am not familiar with moby code. Maybe have a look at https://github.com/moby/moby/blob/master/distribution/push.go to find out? Sorry.

**Comment by jonjohnsonjr** [comment]
The behavior is defined here: https://docs.docker.com/registry/spec/api/#api-version-check

> If 404 Not Found response status, or other unexpected status, is returned, the client should proceed with the assumption that the registry does not implement V2 of the API.

You should return either 401 or 200, depending on your access control setup.

**Comment by majewsky** [comment]
Yes, I'm doing so.

**Comment by jonjohnsonjr** [comment]
I see. Apologies, I had skimmed and misread.

> The spec does not allow the client to just make up scopes

What do you mean by this? What do you expect the behavior to be here?

Some relevant discussion here: https://github.com/containerd/containerd/issues/1318#issuecomment-327277368

tl;dr there are situations where the registry can't possibly know what scopes are required for the client (the operation is ambiguous), but the client *does* know, so it should synthesize them appropriately.

> I suggest that the spec be amended to reflect what Docker actually does

I think the spec should probably be amended (it felt rather under-specified when I implemented it), but there are many clients other than docker, so I don't think "exactly what docker does" is the right end-state.
